rust/azure_iot_operations_connector/src/base_connector/managed_azure_device_registry.rs

```diff
@@ -795,132 +920,64 @@ impl AssetClient {
-                                    log::warn!(
-                                        "Update received for dataset {} on asset {:?}, but DatasetClient has been dropped",
-                                        e_dataset_definition.name,
-                                        self.asset_ref
-                                    );
-                                });
+                    let connector_context_clone = self.connector_context.clone();
+                    let asset_ref_clone = self.asset_ref.clone();
+                    // Spawn a new task to prevent a possible cancellation and ensure the deleted
+                    // notification reaches the application.
+                    tokio::task::spawn(
+                        async move {
+                            Self::unobserve_asset(&connector_context_clone, &asset_ref_clone).await;
                         }
-                    }
-                    // it needs to be created
-                    else if let Err(e) = self.setup_new_dataset(
-                        received_dataset_definition.clone(),
-                        &default_dataset_destinations,
-                    ) {
-                        // If an error is returned, continue to process other datasets even if one isn't valid. Don't give this one to
-                        // the application since we can't forward data on it. If there's an update to the
-                        // definition, they'll get the create notification for it at that point if it's valid
-                        dataset_config_errors.push(e);
-                    }
+                    );
+                    // Asset update has been fully processed, mark as seen.
+                    self.asset_update_watcher_rx.mark_unchanged();
+                    ClientNotification::Deleted
                 }
-
-                // if there were any config errors on the datasets, report them to the ADR service
-                self.report_dataset_config_errors(
-                    dataset_config_errors,
-                    updated_asset.version,
-                    "AssetClient::recv_notification dataset_destination",
-                ).await;
-
-                // update specification
-                let mut unlocked_specification = self.specification.write().unwrap(); // unwrap can't fail unless lock is poisoned
-                *unlocked_specification = AssetSpecification::from(updated_asset);
-
-                ClientNotification::Updated
             },
+            // An error here is not possible since the asset client holds the sender end
+            Ok(()) = self.asset_update_watcher_rx.changed() => {
+                // We need to mark as changed in case `handle_update` is cancelled so we can enter
+                // this branch again
+                self.asset_update_watcher_rx.mark_changed();
+                let updated_asset = self.asset_update_watcher_rx.borrow().clone();
```

Comment:
annoying that this adds 3 extra clones on the updated asset (I'd expect minimum 1 extra to save in the channel), but I don't think there's anything we can do to reduce that, right?

Reply:
For the three cases:
- When sending (`self.asset_update_observation.recv_notification()` branch) we have to clone, it is necessary to save the value as we received it. I could avoid the clone by sending inside of `handle_update` but I would have to send the modified `updated_asset` in there.
- In (`self.asset_update_watcher_rx.changed()` here) I have to clone it because of a similar reason to the above, the updated_asset gets modified and I can't do operations on it on a reference.
- The third case is to save initially correct? I have to do a copy there as well because of the `from` on line 664. (`let specification = AssetSpecification::from(asset);`)